File: recursion/python/solutions/dedupe.py
"""Solution for dedupe."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def dedupe(lst):
    """
    Remove CONSECUTIVE duplicates.

    dedupe(1 -> 1 -> 2 -> 2 -> 1) => 1 -> 2 -> 1
    dedupe(1 -> 2 -> 3) => 1 -> 2 -> 3
    dedupe(()) => ()
    """
    if is_empty(lst):
        return EMPTY_LIST

    if is_empty(lst.next):
        return lst

    first, rest = unprepend(lst)
    second = rest.value

    if first == second:
        return dedupe(rest)
    else:
        return prepend(first, dedupe(rest))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(dedupe(Node.from_list([1, 1, 2, 2, 1]))) == [1, 2, 1]
    assert to_python_list(dedupe(Node.from_list([1, 2, 3]))) == [1, 2, 3]
    assert to_python_list(dedupe(EMPTY_LIST)) == []
    assert to_python_list(dedupe(Node.from_list([1, 1, 1, 1]))) == [1]
    assert to_python_list(dedupe(Node.from_list([1]))) == [1]

    print("All tests passed!")

File: recursion/javascript/solutions/dedupe.js
/**
 * Solution for dedupe.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function dedupe(list) {
  if (isEmpty(list)) {
    return EMPTY_LIST;
  }

  if (isEmpty(list.next)) {
    return list;
  }

  let [first, rest] = unprepend(list);
  let second = rest.value;

  if (first === second) {
    return dedupe(rest);
  }
  return prepend(first, dedupe(rest));
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(dedupe(Node.fromArray([1, 1, 2, 2, 1])))) === '[1,2,1]');
  console.assert(JSON.stringify(toArray(dedupe(Node.fromArray([1, 2, 3])))) === '[1,2,3]');
  console.assert(JSON.stringify(toArray(dedupe(Node.fromArray([1, 1, 1])))) === '[1]');
  console.assert(JSON.stringify(toArray(dedupe(EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { dedupe };
